File: services/server/src/main/java/com/yaytsa/server/error/GlobalExceptionHandler.java
package com.yaytsa.server.error;

import jakarta.servlet.http.HttpServletRequest;
import java.time.Instant;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.slf4j.MDC;
import org.springframework.http.HttpStatus;
import org.springframework.http.ProblemDetail;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.AccessDeniedException;
import org.springframework.security.core.AuthenticationException;
import org.springframework.web.bind.MissingServletRequestParameterException;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.RestControllerAdvice;
import org.springframework.web.method.annotation.MethodArgumentTypeMismatchException;
import org.springframework.web.server.ResponseStatusException;
import org.springframework.web.servlet.resource.NoResourceFoundException;

@RestControllerAdvice
public class GlobalExceptionHandler {

  private static final Logger log = LoggerFactory.getLogger(GlobalExceptionHandler.class);

  @ExceptionHandler(ResponseStatusException.class)
  public ResponseEntity<ProblemDetail> handleResponseStatusException(
      ResponseStatusException ex, HttpServletRequest request) {
    HttpStatus status = HttpStatus.valueOf(ex.getStatusCode().value());
    ProblemDetail problem = createProblemDetail(status, ex.getReason(), request);

    if (status.is5xxServerError()) {
      log.error(
          "ResponseStatusException: status={}, path={}, reason={}",
          status.value(),
          request.getRequestURI(),
          ex.getReason(),
          ex);
    } else if (status == HttpStatus.FORBIDDEN) {
      log.warn(
          "Access denied: status={}, path={}, reason={}",
          status.value(),
          request.getRequestURI(),
          ex.getReason());
    } else {
      log.debug(
          "ResponseStatusException: status={}, path={}, reason={}",
          status.value(),
          request.getRequestURI(),
          ex.getReason());
    }

    return ResponseEntity.status(status).body(problem);
  }

  @ExceptionHandler(IllegalArgumentException.class)
  public ResponseEntity<ProblemDetail> handleIllegalArgument(
      IllegalArgumentException ex, HttpServletRequest request) {
    ProblemDetail problem = createProblemDetail(HttpStatus.BAD_REQUEST, ex.getMessage(), request);

    log.warn("Bad request: path={}, error={}", request.getRequestURI(), ex.getMessage());

    return ResponseEntity.badRequest().body(problem);
  }

  @ExceptionHandler(MethodArgumentTypeMismatchException.class)
  public ResponseEntity<ProblemDetail> handleTypeMismatch(
      MethodArgumentTypeMismatchException ex, HttpServletRequest request) {
    String message =
        String.format("Invalid value '%s' for parameter '%s'", ex.getValue(), ex.getName());
    ProblemDetail problem = createProblemDetail(HttpStatus.BAD_REQUEST, message, request);

    log.warn(
        "Type mismatch: path={}, param={}, value={}",
        request.getRequestURI(),
        ex.getName(),
        ex.getValue());

    return ResponseEntity.badRequest().body(problem);
  }

  @ExceptionHandler(MissingServletRequestParameterException.class)
  public ResponseEntity<ProblemDetail> handleMissingParam(
      MissingServletRequestParameterException ex, HttpServletRequest request) {
    String message = String.format("Missing required parameter: %s", ex.getParameterName());
    ProblemDetail problem = createProblemDetail(HttpStatus.BAD_REQUEST, message, request);

    log.warn(
        "Missing parameter: path={}, param={}", request.getRequestURI(), ex.getParameterName());

    return ResponseEntity.badRequest().body(problem);
  }

  @ExceptionHandler(NoResourceFoundException.class)
  public ResponseEntity<ProblemDetail> handleNoResourceFound(
      NoResourceFoundException ex, HttpServletRequest request) {
    ProblemDetail problem =
        createProblemDetail(HttpStatus.NOT_FOUND, "Resource not found", request);

    log.debug("Resource not found: path={}", request.getRequestURI());

    return ResponseEntity.status(HttpStatus.NOT_FOUND).body(problem);
  }

  @ExceptionHandler(AuthenticationException.class)
  public ResponseEntity<ProblemDetail> handleAuthenticationException(
      AuthenticationException ex, HttpServletRequest request) {
    ProblemDetail problem =
        createProblemDetail(HttpStatus.UNAUTHORIZED, "Authentication required", request);

    log.warn("Authentication failed: path={}", request.getRequestURI());

    return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body(problem);
  }

  @ExceptionHandler(AccessDeniedException.class)
  public ResponseEntity<ProblemDetail> handleAccessDenied(
      AccessDeniedException ex, HttpServletRequest request) {
    ProblemDetail problem = createProblemDetail(HttpStatus.FORBIDDEN, "Access denied", request);

    log.warn("Access denied: path={}, user={}", request.getRequestURI(), MDC.get("userId"));

    return ResponseEntity.status(HttpStatus.FORBIDDEN).body(problem);
  }

  @ExceptionHandler(Exception.class)
  public ResponseEntity<ProblemDetail> handleGenericException(
      Exception ex, HttpServletRequest request) {
    ProblemDetail problem =
        createProblemDetail(
            HttpStatus.INTERNAL_SERVER_ERROR, "An unexpected error occurred", request);

    log.error(
        "Unhandled exception: path={}, type={}, message={}",
        request.getRequestURI(),
        ex.getClass().getSimpleName(),
        ex.getMessage(),
        ex);

    return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(problem);
  }

  private ProblemDetail createProblemDetail(
      HttpStatus status, String detail, HttpServletRequest request) {
    ProblemDetail problem = ProblemDetail.forStatusAndDetail(status, detail);
    problem.setTitle(status.getReasonPhrase());
    problem.setProperty("path", request.getRequestURI());
    problem.setProperty("timestamp", Instant.now().toString());

    String requestId = MDC.get("requestId");
    if (requestId != null) {
      problem.setProperty("requestId", requestId);
    }

    return problem;
  }
}

File: app/src/main/kotlin/dev/yaytsa/app/GlobalExceptionHandler.kt
package dev.yaytsa.app

import org.slf4j.LoggerFactory
import org.springframework.dao.DataAccessException
import org.springframework.dao.InvalidDataAccessApiUsageException
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.http.converter.HttpMessageNotReadableException
import org.springframework.security.access.AccessDeniedException
import org.springframework.web.HttpRequestMethodNotSupportedException
import org.springframework.web.bind.MissingServletRequestParameterException
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.RestControllerAdvice
import org.springframework.web.method.annotation.MethodArgumentTypeMismatchException
import org.springframework.web.server.ResponseStatusException
import org.springframework.web.servlet.NoHandlerFoundException

@RestControllerAdvice
class GlobalExceptionHandler {
    private val log = LoggerFactory.getLogger(GlobalExceptionHandler::class.java)

    @ExceptionHandler(NoHandlerFoundException::class)
    fun handleNotFound(ex: NoHandlerFoundException): ResponseEntity<Map<String, Any>> = errorResponse(HttpStatus.NOT_FOUND, "Not found: ${ex.requestURL}")

    @ExceptionHandler(AccessDeniedException::class)
    fun handleAccessDenied(ex: AccessDeniedException): ResponseEntity<Map<String, Any>> = errorResponse(HttpStatus.FORBIDDEN, "Access denied")

    @ExceptionHandler(IllegalArgumentException::class)
    fun handleBadRequest(ex: IllegalArgumentException): ResponseEntity<Map<String, Any>> {
        log.debug("Bad request: {}", ex.message)
        return errorResponse(HttpStatus.BAD_REQUEST, ex.message ?: "Bad request")
    }

    @ExceptionHandler(HttpMessageNotReadableException::class)
    fun handleUnreadable(ex: HttpMessageNotReadableException): ResponseEntity<Map<String, Any>> = errorResponse(HttpStatus.BAD_REQUEST, "Invalid request body")

    @ExceptionHandler(MissingServletRequestParameterException::class)
    fun handleMissingParam(ex: MissingServletRequestParameterException): ResponseEntity<Map<String, Any>> =
        errorResponse(HttpStatus.BAD_REQUEST, "Missing required parameter: ${ex.parameterName}")

    @ExceptionHandler(HttpRequestMethodNotSupportedException::class)
    fun handleMethodNotAllowed(ex: HttpRequestMethodNotSupportedException): ResponseEntity<Map<String, Any>> =
        errorResponse(HttpStatus.METHOD_NOT_ALLOWED, "Method ${ex.method} not allowed")

    @ExceptionHandler(InvalidDataAccessApiUsageException::class)
    fun handleInvalidDataAccess(ex: InvalidDataAccessApiUsageException): ResponseEntity<Map<String, Any>> {
        log.debug("Invalid data access: {}", ex.message)
        val message = ex.cause?.message ?: ex.message ?: "Invalid request data"
        return errorResponse(HttpStatus.BAD_REQUEST, message)
    }

    @ExceptionHandler(MethodArgumentTypeMismatchException::class)
    fun handleTypeMismatch(ex: MethodArgumentTypeMismatchException): ResponseEntity<Map<String, Any>> {
        log.debug("Type mismatch for parameter '{}': {}", ex.name, ex.message)
        return errorResponse(HttpStatus.BAD_REQUEST, "Invalid value for parameter '${ex.name}': ${ex.value}")
    }

    @ExceptionHandler(DataAccessException::class)
    fun handleDataAccess(ex: DataAccessException): ResponseEntity<Map<String, Any>> {
        log.debug("Data access error: {}", ex.message)
        return errorResponse(HttpStatus.BAD_REQUEST, "Invalid request data")
    }

    @ExceptionHandler(ResponseStatusException::class)
    fun handleResponseStatus(ex: ResponseStatusException): ResponseEntity<Map<String, Any>> =
        errorResponse(HttpStatus.valueOf(ex.statusCode.value()), ex.reason ?: "Error")

    @ExceptionHandler(Exception::class)
    fun handleGeneric(ex: Exception): ResponseEntity<Map<String, Any>> {
        log.error("Unhandled exception", ex)
        return errorResponse(HttpStatus.INTERNAL_SERVER_ERROR, "Internal server error")
    }

    private fun errorResponse(
        status: HttpStatus,
        message: String,
    ): ResponseEntity<Map<String, Any>> =
        ResponseEntity.status(status).body(
            mapOf(
                "error" to message,
                "status" to status.value(),
            ),
        )
}
